### PR TITLE
Maps navigation entry

### PR DESCRIPTION
**Pull request by jakobroehrl**
https://github.com/nextcloud/photos/issues/8
Is this going into the right direction?

Right now I get an error: 
Undefined class constant 'MAPS'

### CODE REVIEW

File: lib/Controller/PageController.php [hunk 1]
```diff
@@ -68,12 +68,12 @@ public function index(): TemplateResponse {
 		$this->eventDispatcher->dispatch(LoadViewer::class, new LoadViewer());
 
 		$this->initialStateService->provideInitialState($this->appName, 'mimes', Application::MIMES);
+		$this->initialStateService->provideInitialState($this->appName, 'maps', Application::MAPS);
```
Comment (skjnldsv) [inline lib/Controller/PageController.php]: `Application::MAPS` doesn't exists,
you need to fetch the proper data for it.
You can check if an app is enabled by using this
`$this->config->getAppValue('maps', 'enabled', 'no') === 'yes')`
Make sure you inject the config dependency in this file's constructor as well

File: src/Photos.vue [hunk 2]
```diff
@@ -22,6 +22,7 @@
 
 <template>
 	<Content app-name="photos">
+		{{MapsInstalled}}
```
Comment (skjnldsv) [inline src/Photos.vue]: you cannot use variables that are not in computed/data in the template.
Also please use variables name like `isMapsInstalled` as it's more straightforward :)

File: lib/Controller/PageController.php (hunk 1 above)
Comment (jakobroehrl) [reply lib/Controller/PageController.php]: Sorry, no idea what to do here. 
I don't know where to add "$this->config->getAppValue('maps', 'enabled', 'no') === 'yes')" 
and what "Make sure you inject the config dependency in this file's constructor as well" could mean...

Comment (skjnldsv) [reply lib/Controller/PageController.php]: Are you familiar with our nextcloud app ecosystem? :)
We updated our tutorial recently here: https://docs.nextcloud.com/server/stable/developer_manual/app/tutorial.html

the provideInitialState requires 3 argments, the name of the app, the unique key of the data set and the data you want to pass.
Here we want to pass if the maps is enabled or not (boolean).
`Application::MAPS` doesn't exists. Look at the Application.php file.
https://github.com/nextcloud/photos/blob/master/lib/AppInfo/Application.php

Comment (skjnldsv) [reply lib/Controller/PageController.php]: So you need to pass the data we want, meaning `$this->config->getAppValue('maps', 'enabled', 'no') === 'yes')` that will return true/false if maps is enabled.

BUT $this->config is not injected here
https://github.com/nextcloud/photos/blob/c73e919829fd43d0f133609a8457790d447baa13/lib/Controller/PageController.php#L47-L55
You need to import [the IConfig interface](https://docs.nextcloud.com/server/stable/developer_manual/api/OCP/IConfig.html) the same way we inject the IRequest or IEventDispatcher ;)

Then you will be able to pass `$this->config->getAppValue('maps', 'enabled', 'no') === 'yes')` as the maps initial state

Comment (skjnldsv) [reply lib/Controller/PageController.php]: ANd the nextcloud php ecosystem will do the rest and give you access to `$config`, which you can then assign to `$this->config` and use it :)

File: src/Photos.vue (hunk 2 above)
Comment (skjnldsv) [reply src/Photos.vue]: @jakobroehrl ^

**Comment by jakobroehrl** [comment]
Thanks for you detailed description. Understand a little bit more :)
So, now I got the state, but have problems to display it:
Uncaught Error: Could not find initial state undefined of maps
    at loadState (index.js:14)
    at Module../src/services/MapsInstalled.js (MapsInstalled.js:24)

File: lib/Controller/PageController.php (hunk 1 above)
Comment (skjnldsv) [reply lib/Controller/PageController.php]: Well done :)

File: src/Photos.vue (hunk 2 above)
Comment (jakobroehrl) [reply src/Photos.vue]: will change it, when it's working. 
There is still one error:
Uncaught Error: Could not find initial state undefined of maps
at loadState (index.js:14)
at Module../src/services/MapsInstalled.js (MapsInstalled.js:24)

File: lib/Controller/PageController.php [hunk 3]
```diff
@@ -47,12 +48,14 @@ class PageController extends Controller {
 	public function __construct($appName,
 								IRequest $request,
 								IEventDispatcher $eventDispatcher,
+								IConfig $config,
 								IInitialStateService $initialStateService) {
-		parent::__construct($appName, $request);
+		parent::__construct($appName, $request, $config);
```
Comment (skjnldsv) [inline lib/Controller/PageController.php]: This is not needed here. :)
The construct of the parent (Controller, see line 38) doesn't needs $config

Comment (skjnldsv) [reply lib/Controller/PageController.php]: https://github.com/nextcloud/server/blob/master/lib/public/AppFramework/Controller.php#L74

File: src/Photos.vue (hunk 2 above)
Comment (skjnldsv) [reply src/Photos.vue]: loadstate() require the app name as first argument https://github.com/nextcloud/photos/blob/master/src/services/AllowedMimes.js#L25

Comment (jakobroehrl) [reply src/Photos.vue]: > 
> 
> loadstate() require the app name as first argument https://github.com/nextcloud/photos/blob/master/src/services/AllowedMimes.js#L25

Found this 2mins before by myself, but it doesn't work neither

Comment (skjnldsv) [reply src/Photos.vue]: It works for me :)

Comment (jakobroehrl) [reply src/Photos.vue]: OK, now it's working. :)

**Comment by jakobroehrl** [comment]
Last question: How should I change the link here to navigate correctly?

- https://github.com/nextcloud/photos/blob/74120cc0e33eff540104a7912d05d896ab430bcd/src/Photos.vue#L35
- https://github.com/nextcloud/photos/blob/74120cc0e33eff540104a7912d05d896ab430bcd/src/Photos.vue#L74-L80

File: src/Photos.vue [hunk 4]
```diff
@@ -69,10 +68,18 @@ export default {
...
-			MapsInstalled
+			videoplaceholder
 		}
 	},
+	computed: {
+		mapsLink() {
+			if (isMapsInstalled) {
+				return 'apps/maps'
+			} else {
+				return 'https://apps.nextcloud.com/apps/maps'
```
Comment (skjnldsv) [inline src/Photos.vue]: I'm wondering if we should rather redirect to the nextcloud apps settings ?
https://nc/settings/apps/integration/maps

File: src/Photos.vue [hunk 5]
```diff
@@ -67,9 +68,18 @@ export default {
...
 			svgplaceholder,
 			imgplaceholder,
-			videoplaceholder,
+			videoplaceholder
 		}
 	},
+	computed: {
+		mapsLink() {
+			if (isMapsInstalled) {
+				return 'apps/maps'
```
Comment (skjnldsv) [inline src/Photos.vue]: You should return thesame as the original object: `{name: 'maps'}`

Comment (jakobroehrl) [reply src/Photos.vue]: 1. How can I get the individual server path? http(s)://host/path/index.php
2. How do I have to change the router to navigate to the new route? It appends the route path only on the existing route in the browser

Thanks!

File: src/Photos.vue (hunk 4 above)
Comment (jakobroehrl) [reply src/Photos.vue]: changed it

File: src/Photos.vue (hunk 5 above)
Comment (skjnldsv) [reply src/Photos.vue]: you need to use the generateUrl method for both.
Ignore my first comment about `{name: 'maps'}` I got confused about the old code :)
What you did was good, but you need to use the generateUrl.

Comment (skjnldsv) [reply src/Photos.vue]: https://github.com/nextcloud/photos/blob/b2c8850d058551e71fa672e7c0002ef4f02ca65c/src/router/index.js#L23
https://github.com/nextcloud/photos/blob/b2c8850d058551e71fa672e7c0002ef4f02ca65c/src/router/index.js#L37

**Comment by jakobroehrl** [comment]
What do you think about this approach?
https://github.com/nextcloud/photos/blob/903091675cb3e6ba7c4779650fcf82413c8167f5/src/router/index.js#L86-L95

Sill have the problem, there is no redirect, it only appends to the URL:
grafik

File: src/router/index.js [hunk 6]
```diff
@@ -84,8 +84,13 @@ export default new Router({
 			}),
 		},
 		{
-			path: '/maps',
-			name: 'maps',
+			path: generateUrl('apps/maps', ''),
```
Comment (skjnldsv) [inline src/router/index.js]: you can use the generateurl directly in the Photos.vue file, there is no need to create new routes :)

File: src/Photos.vue [hunk 7]
```diff
@@ -70,6 +71,15 @@ export default {
 			videoplaceholder,
 		}
 	},
+	computed: {
+		mapsLink() {
+			if (isMapsInstalled) {
+				return {name: 'MapsInstalled'}
```
Comment (skjnldsv) [inline src/Photos.vue]:
```suggestion
				return generateUrl('apps/maps')
```

File: src/Photos.vue [hunk 8]
```diff
@@ -70,6 +71,15 @@ export default {
...
+			} else {
+				return {name: 'MapsNotInstalled'}
```
Comment (skjnldsv) [inline src/Photos.vue]:
```suggestion
				return generateUrl('/settings/apps/integration/maps')
```

File: src/router/index.js (hunk 6 above)
Comment (skjnldsv) [reply src/router/index.js]: See other comments :)

File: src/Photos.vue (hunk 7 above)
Comment (skjnldsv) [reply src/Photos.vue]: (don't forget to import the generateUrl method on this file too :) )

File: src/router/index.js [hunk 9]
```diff
@@ -84,13 +84,9 @@ export default new Router({
 			}),
 		},
 		{
+			path: '/maps',
```
Comment (skjnldsv) [inline src/router/index.js]: you should be able to also drop this entry :)

**Comment by jakobroehrl** [comment]
Hey @skjnldsv I changed the routes. 
But still have this problem:
https://user-images.githubusercontent.com/24530680/70143349-a27c0480-169b-11ea-8278-739fa32d0e74.png

**Comment by skjnldsv** [comment]
> But still have this problem:
> [user-images.githubusercontent.com/24530680/70143349-a27c0480-169b-11ea-8278-739fa32d0e74.png](https://user-images.githubusercontent.com/24530680/70143349-a27c0480-169b-11ea-8278-739fa32d0e74.png)

Ah right! :/
We indeed do not support external url with vue router
Let me manage this  :)

**Comment by skjnldsv** [comment]
Pushed!
@jakobroehrl please review :)

**Pull request merged**